### PR TITLE
Add new layout() test for projects sharing common code in parent folders

### PR DESCRIPTION
**Pull request by memsharded**
Changelog: Feature: Added a new ``self.folders.subproject`` for ``layout()`` to be able to define layouts that goes up to the parent or siblings folders, together with the ``self.folders.root = ".."``.
Docs: https://github.com/conan-io/docs/pull/2662

### CODE REVIEW

**Review by prince-chrismc — APPROVED** [review]
This works 🚀 

The downside to the custom layout is losing the generator folder being set by `cmake_layout` and it's [handling around the conf and multi](https://github.com/conan-io/conan/commit/005fc53485557b0a570bb71670f2ca9c66082165#diff-dbf5d536275be2e3c8db49e46f4b65b830abcb38becc44a077c7d3271c2e0d82R28-R31)

File: conan/tools/cmake/layout.py [hunk 1]
```diff
@@ -14,13 +14,13 @@ def cmake_layout(conanfile, generator=None, src_folder="."):
         else:
             multi = False
 
-    conanfile.folders.source = src_folder
+    conanfile.folders.source = src_folder if not subfolder else os.path.join(subfolder, src_folder)
```
Comment (prince-chrismc) [inline conan/tools/cmake/layout.py]: sub folder is above source dir? 😕 Is that from the perspective of the root folder?

Maybe "nested" or "project" not sure whats better, maybe some is enough?

Comment (memsharded) [reply conan/tools/cmake/layout.py]: Yes, the idea is that ``subfolder`` is from the "root" folder, to get you back to the subproject folder where the conanfile.py is. And then, the ``src_folder`` defines the source folder within that subproject (typically, where the CMakeLists.txt of that subproject is)

**Review by prince-chrismc — APPROVED** [review]

**Review by lasote — CHANGES_REQUESTED** [review]
I think understanding why going to `..` to later come back to `pkg` is too much. There is no way a "subfolder" argument will be interpreted and used in the correct way.

File: conans/test/functional/layout/test_in_subfolder.py [hunk 2]
```diff
@@ -54,3 +58,112 @@ def build(self):
...
+        from conan.tools.files import load, copy, save
+        class Pkg(ConanFile):
+            name = "pkg"
+            version = "0.1"
+            settings = "os", "compiler", "build_type", "arch"
+            generators = "CMakeToolchain"
+
+            def layout(self):
+                self.folders.root = ".."
+                cmake_layout(self, subfolder="pkg")
```
Comment (memsharded) [inline conans/test/functional/layout/test_in_subfolder.py]:
```suggestion
            def layout(self):
                self.folders.root = ".."
                self.folders.subproject = "pkg"
                cmake_layout(self)
```

Comment (memsharded) [reply conans/test/functional/layout/test_in_subfolder.py]: Done

**Comment by memsharded** [comment]
I have changed the approach, with a ``self.folders.subproject`` definition, and extended its application to other generators as well. Please re-check.

File: conans/model/layout.py [hunk 3]
```diff
@@ -31,6 +31,10 @@ def __init__(self):
         # Relative location of the project root, if the conanfile is not in that project root, but
         # in a subfolder: e.g: If the conanfile is in a subfolder then self.root = ".."
         self.root = None
+        # The relative location wrt to the project root of the subproject containing the
```
Comment (lasote) [inline conans/model/layout.py]:
```suggestion
        # The relative location with respect to the project root of the subproject containing the
```

We hate acronyms

**Review by lasote — APPROVED** [review]

**Review by franramirez688 — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]
👍 subproject is a more clear name

**Pull request merged**